feat(CollectiblesView): Add combobox/popup with custom filtering options

caybro opened the pull request:
.. to filter by community or collection name

- make the HistoryView own filter button look like the other combos
- fix some cosmetic issues for StatusCombo in small/secondary mode
- fix StatusBaseInput bg color in dark mode (was invisible)
- update CollectiblesViewPage with options to include regular and/or
community collectibles

Fixes https://github.com/status-im/status-desktop/issues/12969
Fixes https://github.com/status-im/status-desktop/issues/12948

### What does the PR do

Adds a filtering combo to the collectibles view

### Affected areas

Wallet/CollectiblesView

### Screenshot of functionality (including design for comparison)

- [x] I've checked the design and this PR matches it

![Snímek obrazovky z 2023-12-13 16-00-11](https://github.com/status-im/status-desktop/assets/5377645/bdf5a098-dc33-486f-b263-4070c66ec2e4)

![Snímek obrazovky z 2023-12-13 16-00-55](https://github.com/status-im/status-desktop/assets/5377645/4ff63c87-e6ff-4310-9e20-9b74b6b1ee4e)

Storybook with optional collectibles' sources:
![image](https://github.com/status-im/status-desktop/assets/5377645/3e5cc62c-53e7-4ea5-9e92-5e5d80acd20c)

caybro commented on ui/app/AppLayouts/Wallet/controls/FilterComboBox.qml:
```diff
@@ -0,0 +1,319 @@
+ComboBox {
+    id: root
+
+    required property var controller
```
This should probably be just the 2 models instead...

caybro replied:
Fixed

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 6e4682ae | [#1](https://ci.status.im/job/status-desktop/job/prs/job/linux/job/x86_64/job/tests-nim/job/PR-13024/1/) | 2023-12-18 10:24:01 | ~6 min | `tests/nim` | [:page_facing_up:`log`](https://ci.status.im/job/status-desktop/job/prs/job/linux/job/x86_64/job/tests-nim/job/PR-13024/1//consoleText) |

noeliaSD commented on ui/app/AppLayouts/Wallet/controls/FilterComboBox.qml:
Where is it the `storybook` page?? :wink:

caybro replied:
It's in CollectiblesViewPage; imho it doesn't make sense to have a standalone page for this, it needs some view with different collectibles

noeliaSD commented on ui/app/AppLayouts/Wallet/controls/FilterComboBox.qml:
```diff
@@ -0,0 +1,319 @@
+ComboBox {
+    id: root
+
+    required property var regularTokensModel
```
Will these be the ones that are NOT part of any collection? Maybe a comment could help to clarify!

caybro replied:
Yup, will add comments to what is what

caybro replied:
Fixed

noeliaSD commented on ui/app/AppLayouts/Wallet/controls/FilterComboBox.qml:
```diff
@@ -0,0 +1,319 @@
+
+    displayText: qsTr("Collection")
+
+    horizontalPadding: 12
```
Some `Style.current.value`?!

caybro replied:
There's none for `12`

noeliaSD reviewed (changes requested):
From the images shared in the PR it looks nice!! But I could not test the new combobox added since there's no `storybook` page for it (it'd be nice to add it ;) ) to test isolated and I experience a crash when trying to open `Wallet main layout`. More info in the following comments:

- There's the following line that probably should not be present when regular collectibles model is empty:
<img width="890" alt="Screenshot 2023-12-19 at 10 22 47" src="https://github.com/status-im/status-desktop/assets/97019400/a2def59e-0e59-496b-9a3c-c47c02894e28">

- Whenever I enter in `Wallet Main View` there's the following crash (tried on master and it's not reproducible):
```
DBG 2023-12-19 10:39:52.573+01:00 NewBE_callPrivateRPC                       topics="rpc" tid=144102 file=core.nim:27 rpc_method=wakuext_slowdownArchivesImport
DBG 2023-12-19 10:39:52.731+01:00 NewBE_callPrivateRPC                       topics="rpc" tid=144102 file=core.nim:27 rpc_method=accounts_hasPairedDevices
DBG 2023-12-19 10:39:52.765+01:00 NewBE_callPrivateRPC                       topics="rpc" tid=144102 file=core.nim:27 rpc_method=wallet_getOwnedCollectiblesAsync
Warning: void ManageTokensController::loadSettings() Missing symbol while reading tokens settings (:0, )
Warning: qrc:/app/AppLayouts/Wallet/views/AssetsDetailView.qml:32:9: Unable to assign [undefined] to QString (qrc:/app/AppLayouts/Wallet/views/AssetsDetailView.qml:32, )
Traceback (most recent call last)
/status-desktop/src/nim_status_client.nim(208) nim_status_client
/status-desktop/src/nim_status_client.nim(205) mainProc
/status-desktop/vendor/nimqml/src/nimqml/private/qguiapplication.nim(29) exec
/status-desktop/vendor/nimqml/src/nimqml/private/qobject.nim(52) qobjectCallback
/status-desktop/vendor/nimqml/src/nimqml/private/qobject.nim(37) onSlotCalled
/status-desktop/vendor/nimbus-build-system/vendor/Nim/lib/core/macros.nim(558) onSlotCalled
/status-desktop/src/app/core/signals/signals_manager.nim(63) receiveSignal
/status-desktop/src/app/core/signals/signals_manager.nim(60) processSignal
/status-desktop/src/app/core/eventemitter.nim(58) emit
/status-desktop/src/app/modules/shared_modules/collectibles/events_handler.nim(116) :anonymous
/status-desktop/src/app/modules/shared_modules/collectibles/events_handler.nim(69) handleApiEvents
/status-desktop/src/app/modules/shared_modules/collectibles/controller.nim(156) :anonymous
/status-desktop/src/app/modules/shared_modules/collectibles/controller.nim(126) processGetOwnedCollectiblesResponse
/status-desktop/src/app/modules/shared_models/collectibles_model.nim(94) setIsFetching
/status-desktop/src/app/modules/shared_models/collectibles_model.nim(325) checkLoadingItems
/status-desktop/src/app/modules/shared_models/collectibles_model.nim(316) removeLoadingItems
/status-desktop/vendor/nimqml/src/nimqml/private/qabstractitemmodel.nim(196) endRemoveRows
SIGSEGV: Illegal storage access. (Attempt to read from nil?)
```

caybro commented on ui/StatusQ/src/StatusQ/Controls/StatusBaseInput.qml:
```diff
@@ -278,7 +278,7 @@ Item {
         id: background
         anchors.fill: parent
-        color: root.showBackground ? Theme.palette.baseColor2 : "transparent"
+        color: root.showBackground ? Theme.palette.statusAppNavBar.backgroundColor : "transparent"
```
Yeah... I couldn't really find a better suited color :/

alexjba replied:
IMO it looks much better in dark mode. As opposed to `Theme.palette.baseColor2`

<img width="290" alt="Screenshot 2023-12-21 at 15 19 58" src="https://github.com/status-im/status-desktop/assets/47811206/d998b7a6-c3cf-4913-85c7-f2cffad6ebd3">

alexjba replied:
> IMO it looks much better in dark mode. As opposed to `Theme.palette.baseColor2`
> 
> <img alt="Screenshot 2023-12-21 at 15 19 58" width="290" src="https://private-user-images.githubusercontent.com/47811206/292215804-d998b7a6-c3cf-4913-85c7-f2cffad6ebd3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDMxNjUxOTksIm5iZiI6MTcwMzE2NDg5OSwicGF0aCI6Ii80NzgxMTIwNi8yOTIyMTU4MDQtZDk5OGI3YTYtYzNjZi00OTEzLTg1YzctZjJjZmZhZDZlYmQzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjIxVDEzMjEzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMyZGFkNzI4ZDZmOTkyZjUwOTc0ZDJkNDJjMWYxNmUwYzhlMWQ3NmQ1NGJhMGVlZDdmMzRiZTAxNDYyYmNiMzkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.yBUOCCbtcRtYrVVYn356NuWJyUtMovIGiJe_JwsHHDM">

But the colours on the section headers are wrong IMO. If on light theme it's the same colour as the delegate item, I guess it should be the same colour on the dark theme as well.

caybro replied:
Yeah it's really a different color based on the light/dark theme

caybro commented on ui/app/AppLayouts/Wallet/views/CollectiblesView.qml:
```diff
@@ -87,6 +88,16 @@ ColumnLayout {
                     return d.controller.filterAcceptsSymbol(model.symbol) && (customFilter.isCommunity ? !!model.communityId : !model.communityId)
                 }
+            },
+            ExpressionFilter {
```
Yeah, they are mutually exclusive, that's why they have the `enabled` property set for one or the other case; seemed more straightforward to me, the rest of the properties is exactly the same

caybro commented on ui/app/AppLayouts/Wallet/controls/FilterComboBox.qml:
Yeah maybe it's too generic but it's kinda hard to find a nice unique name with hundreds of QML files :) That's why it's inside the "Wallet/controls" subdir, it's not really useful outside its context

caybro commented on ui/StatusQ/src/wallet/managetokenscontroller.cpp:
```diff
@@ -478,6 +485,46 @@ void ManageTokensController::rebuildCommunityTokenGroupsModel()
+    QStringList collectionIds;
+    QList<TokenData> result;
+
+    const auto count = m_regularTokensModel->count();
```
It's initialized right in the CTOR

caybro commented on ui/StatusQ/src/wallet/managetokenscontroller.cpp:
```diff
@@ -478,6 +485,46 @@ void ManageTokensController::rebuildCommunityTokenGroupsModel()
+        }
+    }
+
+    m_regularTokenGroupsModel->clear();
```
Same here, created in CTOR

caybro commented on ui/StatusQ/src/wallet/managetokenscontroller.cpp:
```diff
@@ -345,6 +349,9 @@ void ManageTokensController::parseSourceModel()
     reloadCommunityIds();
     m_communityTokensModel->setCommunityIds(m_communityIds);
 
+    // build collections
```
No, these are really collections; the other one being "community groups" (already existing code)

caybro commented:
> 
> * There's the following line that probably should not be present when regular collectibles model is empty:
> 

Yeah! Will remove it, thanks

> * Whenever I enter in `Wallet Main View` there's the following crash (tried on master and it's not reproducible):
> 

Hmm, I can't really reproduce this :/ Will rebase onto latest master and try again

noeliaSD commented:
After `storybook` inspection:

- Icon color / background looks different than in the dsign. Here the design:
<img width="390" alt="Screenshot 2023-12-19 at 11 38 33" src="https://github.com/status-im/status-desktop/assets/97019400/322c67b2-8903-4374-aa45-8300dc01f3af">

- I see this warning: `status-desktop/ui/app/AppLayouts/Wallet/views/CollectiblesView.qml:280:9: QML Connections: Detected function "onIsFetchingChanged" in Connections element. This is probably intended to be a signal handler but no signal of the target matches the name.`

caybro commented:
> After `storybook` inspection:
> 
> * Icon color / background looks different than in the dsign. Here the design:

Uhm yeah... you're right... it's an icon that's supposed to look like a button; I'll think of something

> * I see this warning: `status-desktop/ui/app/AppLayouts/Wallet/views/CollectiblesView.qml:280:9: QML Connections: Detected function "onIsFetchingChanged" in Connections element. This is probably intended to be a signal handler but no signal of the target matches the name.`

Unrelated, it's because of the mock models in storybook that lack the signals the real models have

caybro commented:
> > * There's the following line that probably should not be present when regular collectibles model is empty:
> 
> Yeah! Will remove it, thanks

Fixed

> > * Whenever I enter in `Wallet Main View` there's the following crash (tried on master and it's not reproducible):
> 
> Hmm, I can't really reproduce this :/ Will rebase onto latest master and try again

Rebased

noeliaSD commented:
> > > * There's the following line that probably should not be present when regular collectibles model is empty:
> > 
> > 
> > Yeah! Will remove it, thanks
> 
> Fixed
> 
> > > * Whenever I enter in `Wallet Main View` there's the following crash (tried on master and it's not reproducible):
> > 
> > 
> > Hmm, I can't really reproduce this :/ Will rebase onto latest master and try again
> 
> Rebased

After rebase I still see exactly the same crash and log. @alexjba, @micieslak , @alexandraB99 could you try?

noeliaSD commented:
Also another detail. The searcher placeholder should be:

<img width="350" alt="Screenshot 2023-12-19 at 14 15 36" src="https://github.com/status-im/status-desktop/assets/97019400/54a41f12-8eca-4f85-adae-6086a553eee2">

And allow the search by name or id.

caybro commented:
> Also another detail. The searcher placeholder should be:
> 
> <img alt="Screenshot 2023-12-19 at 14 15 36" width="350" src="https://private-user-images.githubusercontent.com/97019400/291596817-54a41f12-8eca-4f85-adae-6086a553eee2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI5OTIxMjksIm5iZiI6MTcwMjk5MTgyOSwicGF0aCI6Ii85NzAxOTQwMC8yOTE1OTY4MTctNTRhNDFmMTItOGVjYS00Zjg1LWFkYWUtNjA4NmE1NTNlZWUyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjE5VDEzMTcwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIwOGY4YmZlMGUxNWZjNzRhMzJmNGRlMjdmNDIyZDI0Njg3NzdjYjU4OWYwNjUwMDRlZGE1YmQyYjIxZGE4ZjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.iOSHUiLsWZxCingr2INHmK5sZIprbaQyuJ2bhsT30KY">
> And allow the search by name or id. [Design](https://www.figma.com/file/eM26pyHZUeAwMLviaS1KJn/%E2%9A%99%EF%B8%8F-Wallet-Settings%3A-Manage-Tokens?type=design&node-id=71%3A103480&mode=design&t=HBhmgzvmu47bpf8o-1)

Right, I'll extend the search to include the collection ID as well

caybro commented:
NB: the `e2e` testsuite is failing because of a crash in `ConcatModel`

caybro commented:
> NB: the `e2e` testsuite is failing because of a crash in `ConcatModel`

Fixed now

caybro commented:
> > Also another detail. The searcher placeholder should be:
> > <img alt="Screenshot 2023-12-19 at 14 15 36" width="350" src="https://private-user-images.githubusercontent.com/97019400/291596817-54a41f12-8eca-4f85-adae-6086a553eee2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI5OTIxMjksIm5iZiI6MTcwMjk5MTgyOSwicGF0aCI6Ii85NzAxOTQwMC8yOTE1OTY4MTctNTRhNDFmMTItOGVjYS00Zjg1LWFkYWUtNjA4NmE1NTNlZWUyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjE5VDEzMTcwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIwOGY4YmZlMGUxNWZjNzRhMzJmNGRlMjdmNDIyZDI0Njg3NzdjYjU4OWYwNjUwMDRlZGE1YmQyYjIxZGE4ZjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.iOSHUiLsWZxCingr2INHmK5sZIprbaQyuJ2bhsT30KY">
> > And allow the search by name or id. [Design](https://www.figma.com/file/eM26pyHZUeAwMLviaS1KJn/%E2%9A%99%EF%B8%8F-Wallet-Settings%3A-Manage-Tokens?type=design&node-id=71%3A103480&mode=design&t=HBhmgzvmu47bpf8o-1)
> 
> Right, I'll extend the search to include the collection ID as well

FTR, this has been implemented as well; the search now also include the ID

alexjba reviewed (commented):
I'll leave this partial review and continue with the code.

I was wondering why the checkboxes are grey in the design. But I think I got it now. The default state should be with all the checkboxes checked and marked by using a grey background (see the screenshot below). When the first checkbox gets checked by the user the colour changes to blue and all the other checkboxes are unchecked. We're missing this flow.

<img width="390" alt="Screenshot 2023-12-21 at 14 58 57" src="https://github.com/status-im/status-desktop/assets/47811206/85c6f2b9-64bd-40c5-85d0-8cb5972aeb2e">

The header is flickering when an item is selected/deselected

https://github.com/status-im/status-desktop/assets/47811206/963048df-7344-4f71-9315-a24fae04c0f5

The middle section drops a shadow when the top section drops it. The shadow should be decoupled.

<img width="295" alt="Screenshot 2023-12-21 at 15 24 02" src="https://github.com/status-im/status-desktop/assets/47811206/888b60f6-ed3b-4e27-86b3-7b1512929350">

Should the `Collection` button change its state while searching?

https://github.com/status-im/status-desktop/assets/47811206/1fe25f0c-c8bc-4372-b9d2-0f2fc367c6bd

caybro commented:
> I'll leave this partial review and continue with the code.
> 
> I was wondering why the checkboxes are grey in the design. But I think I got it now. The default state should be with all the checkboxes checked and marked by using a grey background (see the screenshot below). When the first checkbox gets checked by the user the colour changes to blue and all the other checkboxes are unchecked. We're missing this flow.

Ah, only noticed now! OK, will have to redo the logic a bit there :/ Good catch!

> The header is flickering when an item is selected/deselected

Yup, will fix, probably the OpacityAnimation there
 
> The middle section drops a shadow when the top section drops it. The shadow should be decoupled.

Nice catch too, never noticed

> Should the `Collection` button change its state while searching?

Nope, it should only when the popup is closed, gonna fix

alexjba commented:
> > I'll leave this partial review and continue with the code.
> > I was wondering why the checkboxes are grey in the design. But I think I got it now. The default state should be with all the checkboxes checked and marked by using a grey background (see the screenshot below). When the first checkbox gets checked by the user the colour changes to blue and all the other checkboxes are unchecked. We're missing this flow.
> 
> Ah, only noticed now! OK, will have to redo the logic a bit there :/ Good catch!

I think the `Clear filter` button needs to be considered as well in the new logic. Once everything is checked it's kind of redundant because it will uncheck everything, but the filtering doesn't change.

caybro commented:
> > > I'll leave this partial review and continue with the code.
> > > I was wondering why the checkboxes are grey in the design. But I think I got it now. The default state should be with all the checkboxes checked and marked by using a grey background (see the screenshot below). When the first checkbox gets checked by the user the colour changes to blue and all the other checkboxes are unchecked. We're missing this flow.
> > 
> > 
> > Ah, only noticed now! OK, will have to redo the logic a bit there :/ Good catch!
> 
> I think the `Clear filter` button needs to be considered as well in the new logic. Once everything is checked it's kind of redundant because it will uncheck everything, but the filtering doesn't change.

Yeah, that was my initial idea too ... but then I figured out it might be enough to just change the appearance of the checkbox itself, not the logic :)

caybro commented:
> > > > I'll leave this partial review and continue with the code.
> > > > I was wondering why the checkboxes are grey in the design. But I think I got it now. The default state should be with all the checkboxes checked and marked by using a grey background (see the screenshot below). When the first checkbox gets checked by the user the colour changes to blue and all the other checkboxes are unchecked. We're missing this flow.
> > > 
> > > 
> > > Ah, only noticed now! OK, will have to redo the logic a bit there :/ Good catch!
> > 
> > 
> > I think the `Clear filter` button needs to be considered as well in the new logic. Once everything is checked it's kind of redundant because it will uncheck everything, but the filtering doesn't change.
> 
> Yeah, that was my initial idea too ... but then I figured out it might be enough to just change the appearance of the checkbox itself, not the logic :)

Fixed as well:
[Checkboxes](https://github.com/status-im/status-desktop/assets/5377645/66c87440-6ec1-4e91-a6fe-3b6e04883f2c)

Khushboo-dev-cpp reviewed (approved):
Have just had a brief look at the PR and looks good to me.
